src/rpc/common/Validators.h

```diff
@@ -0,0 +1,247 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2023, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+#include <rpc/common/Concepts.h>
+#include <rpc/common/Specs.h>
+#include <rpc/common/Types.h>
+
+namespace RPCng::validation {
+
+/**
+ * @brief Check that the type is the same as what was expected
+ *
+ * @tparam Expected The expected type that value should be convertible to
+ * @param value The json value to check the type of
+ * @return true if convertible; false otherwise
+ */
+template <typename Expected>
+[[nodiscard]] bool static checkType(boost::json::value const& value)
+{
+    auto has_error = false;
+    if constexpr (std::is_same_v<Expected, bool>)
+    {
+        if (not value.is_bool())
+            has_error = true;
+    }
+    else if constexpr (std::is_same_v<Expected, std::string>)
+    {
+        if (not value.is_string())
+            has_error = true;
+    }
+    else if constexpr (
+        std::is_same_v<Expected, double> or std::is_same_v<Expected, float>)
+    {
+        if (not value.is_double())
+            has_error = true;
+    }
+    else if constexpr (
+        std::is_convertible_v<Expected, uint64_t> or
+        std::is_convertible_v<Expected, int64_t>)
+    {
+        if (not value.is_int64() && not value.is_uint64())
+            has_error = true;
+    }
+    else if constexpr (std::is_same_v<Expected, boost::json::array>)
+    {
+        if (not value.is_array())
+            has_error = true;
+    }
+
+    return not has_error;
+}
+
+/**
+ * @brief A meta-validator that acts as a spec for a sub-object/section
+ */
+class section final
+{
+    std::vector<FieldSpec> specs;
+
+public:
+    explicit section(std::initializer_list<FieldSpec> specs) : specs{specs}
+    {
+    }
+
+    [[nodiscard]] MaybeError
+    verify(boost::json::value const& value, std::string_view key) const;
+};
+
+/**
+ * @brief A validator that simply requires a field to be present
+ */
+struct required final
```

Comment:
"required" is a keyword in C++20. maybe we shall avoid using it as struct name. https://en.cppreference.com/w/cpp/keyword/requires

Reply:
"required" is not a keyword, "requires" is. But you got a point there - they are pretty similar. I was thinking whether we should stick to our convention of `CamelCase` for structs/classes in this case too. What do you think? it would make the spec code look a little different tho.
Another thing I had in mind is, instead of `verify` we could just use `operator()()` and that should allow us to instantly use a lambda or any callable as a `Requirement`. I think. But we can also leave that for the next iteration - clearly this code will still be evolving in the near future.

Reply:
`verify` is more readable

Reply:
that is true. let's leave as is for now at least.